README.md
Mirage
======
Mirage aids testing of your applications by hosting mock responses so that your applications do not have to talk to real endpoints. It's accessible via HTTP and has a RESTful interface.    

Any ideas/improvements or feedback you have are greatly appreciated.  
  
Information on how to use Mirage can be found [here](https://github.com/lashd/mirage/wiki).  
  
I hope you find it useful,  

Leon

P.s. Mirage runs on Linux,MacOSX and Windows; Rubies 1.9.3+ and JRuby.

Installation
------------
    gem install mirage 
    
What's New?
-----------
### 3.0.0
------------------------------
3.0.0 is now out the following is a description of what's new.

A full description of the functionality is also available in the projects [feature files](https://www.relishapp.com/lashd/mirage/docs) hosted on Relish. A lot of effort has gone in to trying to make these tests readable so if something is missing or unclear drop me a line.


#### What's new in the Server:
##### 1: Mirage uses JSON as its communucations medium
Mirage is now configured using JSON. JSON is also used as the output format for Mirage's other operations.
##### 2: Wildcards are now supported in the URI.
You can now specify wild cards in the URI matcher. This means that a url such as '/*/world' would match '/hello/world' :)
##### 3: Full Request Data now tracked
You can now retrieve all data associated with a request that triggers a response. Previously only the the request body/query string was tracked.
Now the full request, including HTTP headers are returned when querying '/requests/template_id'

##### 4. Parameters and body content matchers are now specified seperately
Now you can specify as many parameter and body matchers as you want. These can be both litteral strings or Regex's
  
Previously, it was only possible to specify a single matcher and this would be applied against both the querystring and the request body.
##### 5. HTTP Header matchers
You can now also specify matchers against HTTP headers.
##### 6. More advanced template resolution.
Templates are now scored to find the most appropriate template when finding a match for a particular request. Previously the first potential match was returned even
if there was a more appropriate template.

Litteral matchers are worth more in the scoring process than regex based ones for example.
##### 7. The url has changed
Mirage is now accessible via: http://localhost:7001. I.e. 'mirage' has been removed from all resources 

e.g. responses are now under http://localhost:7001/responses
##### 8. Default responses directory renamed to mirage
Default templates used to be stored in a directory named 'responses'. The term 'response' was the incorrect word to describe what is now called a 'template'. Rather than rename this directory to templates, which is a directory name that may already be in use in your project, the default name for this directory has been changed to 'mirage'.  
  
The old directory name can still be used if you use the client or command line interface to overide where the responses are loaded from

#### What's new in the Client:
##### 1. Template Models
Perhaps the biggest addition to the client. Simply mixin Mirage::Template::Model in to a class, give it a method called body and there you have it... a class that can be used to put objects straight on to Mirage.
 
* All the methods you find on a standard response can then be used at the class level to set defaults. 
 
* Add builder methods to your class using the builder_methods class method.
 
**Example Usage:** (See rdoc for full details)  

    class UserProfile
      extend Mirage::Template::Model
      
      endpoint '/users'
      http_method :get
      status 200
      content_type 'application/json'
      
      builder_methods :firstname, :lastname, :age
      
      def body
        {firstname: firstname, lastname: lastname, age: age}.to_json
      end
    end
    
    
    Mirage::Client.new.put UserProfile.new.firstname('Joe').lastname('blogs')
    
##### Client interface
The client interface has been overhauled to make it more usable. It supports a couple of different ways of specifying a template
to let to specify templates in the style that best suites your code. The block in the examples is executed in the scope of the template
object itself so any method in it can be called. The template is also passed in to the block if you prefer to to call the methods
on a variable.

**Example Usage:** (See rdoc for full details)  

    mirage = Mirage::Client.new
    mirage.put('/users') do
      status 201
      http_method :put
      body 'response'
      
      required_body_content << 'Abracadabra'
      required_headers['Header'] = 'value'
      
      delay 1.5
    end
    
    # Template Model classes can be customised in exactly the same way
    
    mirage.put UserProfile.new do
      status 404
      required_parameters[:name] = /joe.*/
    end
#### What you don't get
A better UI :) it's on its way though!

### 2.4.0
---------
#### What do I get?
##### 1: Configure your client with defaults for each of your responses
Remove the repetition of setting things like the content-type each to time you put a response on to Mirage

**Example Usage:** (See rdoc for full details)  

    client = Mirage::Client.new do |defaults|
      defaults.content_type = "text/xml"
      defaults.delay = 0.5
    end  
    
    client.put("greeting", "hello") -> template put on to Mirage with a delay of 0.5 seconds and content-type text/xml configured
#### What do I have to do to upgrade?
Nothing, this is a new feature. Let me know if you have any problems :)
    
    


### 2.3.0
---------
#### What do I get?
##### 1: Run more than one instance of Mirage per machine.
##### 2: Programmatic interface for starting and stopping Mirage
The client API now includes the ability to start and stop Mirage locally. No more calling out to the command line.  

**Example Usage:** (See rdoc for full details)  

    mirage_client = Mirage.start :port => 9001 #with out args start mirage on 7001 by default
    Mirage.stop # stops the current running instance as long as only one is running  
    
##### 3: Updated command line interface
Now powered by the mighty Thor, the stop subcommand has been enhanced.  
**Usage:**  

    mirage stop -> As it always did, stops the current instances of Mirage as long as there is only one running instance.  
    mirage stop -p port -> stops Mirage on a given port.    
    mirage stop -p port1, port2... -> stop multiple instances of Mirage.  
    mirage stop -p all -> stop all instances of mirage.  
#### What do I have to do to upgrade?
You shouldn't have to do anything to move from the last version of Mirage to this one. Let me know if something stops working!


HISTORY
v3.0.12 - Merge of [pull request from @asehra](https://github.com/lashd/mirage/pull/17). Raises exception when trying to retrieve request information and request has not been received.
V3.0.11 - Implement waiting function within project and remove dependency on waitforit
V3.0.8 - bug fix for issue 13: https://github.com/lashd/mirage/issues/13 - match Content-Length header
V3.0.7 - bug fix for issue 13: https://github.com/lashd/mirage/issues/13 - match Content-Type header
V3.0.6 - bug fix
V3.0.5 - fixed bug in builder methods
V3.0.1 - updating gem post install message to point people towards the readme due checkout the differences between version 2 and 3 of Mirage
V3.0.0 - Major update see README.md
V2.4.2 - update eventmachine, make sure Mirage.start always returns a client
V2.4.1 - adding rdoc for configure method
V2.4.0 - Adding ability to set defaults
V2.3.0 - Enhanced commandline/programmatic api
V2.2.3 - fix bug where mirage is already running a client calls exit.
V2.2.1/2 - Code tidy up
V2.2.0 - Api for starting and stopping Mirage locally
V2.1.2 - rename status_code to status
V2.1.1 - Set response code and tidy ups
V2.0.16 - minor typo fix for window command window title in windows
V2.0.15 - fix uploading files from client and fix starting mirage on windows
V2.0.12/13/14 - updating gem versions
V2.0.8/9/10/11 - tidy up code, still a lot to do :)
V2.0.7 - fix missing dependency on waitforit gem
V2.0.1/2/3 - tidying up code
V2.0.0 - Converted to restful interface


server/server.rb
require 'sinatra'
require 'helpers'
require 'base64'
module Mirage

  class Server < Sinatra::Base

    REQUESTS = {}

    helpers Helpers::TemplateRequirements, Helpers::HttpHeaders

    put '/templates/*' do |name|
      content_type :json
      mock_response = synchronize do
        MockResponse.new(name, JSON.parse(request.body.read))
      end

      mock_response.requests_url = request.url.gsub("/templates/#{name}", "/requests/#{mock_response.response_id}")
      {:id => mock_response.response_id}.to_json
    end

    %w(get post delete put options).each do |http_method|
      send(http_method, '/responses/*') do |name|
        body, query_string = Rack::Utils.unescape(request.body.read.to_s), request.query_string

          options = {:body => body,
                   :http_method => http_method,
                   :endpoint => name,
                   :params => request.params,
                   :headers => extract_http_headers(env)}
        begin
          record = MockResponse.find(options)
        rescue ServerResponseNotFound
          record = MockResponse.find_default(options)
        end

        synchronize do
          REQUESTS[record.response_id] = request.dup
        end


        send_response(record, body, request, query_string)
      end
    end

    get '/templates/:id/preview' do
      send_response(MockResponse.find_by_id(response_id), '', {}, '')
    end

    delete '/templates/:id' do
      synchronize do
        MockResponse.delete(response_id)
        REQUESTS.delete(response_id)
      end

      200
    end

    delete '/requests' do
      synchronize do
        REQUESTS.clear
      end

      200
    end

    delete '/requests/:id' do
      synchronize do
        REQUESTS.delete(response_id)
      end

      200
    end

    delete '/templates' do
      synchronize do
        REQUESTS.clear
        MockResponse.delete_all
      end

      200
    end

    get '/templates/:id' do
      MockResponse.find_by_id(response_id).raw
    end

    get '/requests/:id' do
      content_type :json
      tracked_request = REQUESTS[response_id]
      if tracked_request

        tracked_request.body.rewind
        body = tracked_request.body.read

        parameters = tracked_request.params.dup.select { |key, value| key != body }

        {id: request.url,
         request_url: tracked_request.url,
         headers: extract_http_headers(tracked_request.env),
         parameters: parameters,
         body: body}.to_json

      else
        404
      end
    end

    get '/' do
      haml :index
    end


    put '/defaults' do
      synchronize do
        MockResponse.delete_all
        if File.directory?(settings.defaults.to_s)
          Dir["#{settings.defaults}/**/*.rb"].each do |default|
            begin
              eval File.read(default)
            rescue Exception => e
              raise "Unable to load default responses from: #{default}"
            end
          end
        end
      end
      200
    end
#
    put '/backup' do
      synchronize do
        MockResponse.backup
      end

      200
    end


    put '/' do
      synchronize do
        MockResponse.revert
      end

      200
    end

    get '/pid' do
      "#{$$}"
    end

    error ServerResponseNotFound do
      404
    end

    error do
      erb request.env['sinatra.error'].message
    end

    helpers do

      def synchronize &block
        Mutex.new.synchronize &block
      end

      def response_id
        params[:id].to_i
      end

      def prime &block
        block.call Mirage::Client.new "http://localhost:#{settings.port}"
      end

      def send_response(mock_response, body='', request={}, query_string='')
        sleep mock_response.response_spec['delay']
        content_type(mock_response.response_spec['content_type'])
        status mock_response.response_spec['status']
        headers mock_response.headers
        mock_response.value(body, request, query_string)
      end


    end
  end
end